crates/core/src/vm.rs

```diff
@@ -76,10 +77,44 @@ pub fn build_query<'a>(
                     .take()
                     .map(Ok)
                     .unwrap_or_else(|| get_table(ctx, stdb, tx, &query.source, sources))?;
-                let header = result.head().clone();
-                let cmp: ColumnOp = index_scan.clone().into();
-                let iter = result.select(move |row| cmp.compare(row, &header));
-                Box::new(iter)
+
+                let cols = &index_scan.columns;
+                let bounds = &index_scan.bounds;
+                if cols.is_singleton() {
+                    // For singleton constraints, we compare the column directly against `bounds`.
+                    let head = cols.head().idx();
+                    let iter = result.select(move |row| Ok(bounds.contains(&*row.read_column(head).unwrap())));
+                    Box::new(iter) as Box<IterRows<'a>>
+                } else {
+                    // For multi-col constraints, these are stored as bounds of product values,
+                    // so we need to project these into single-col bounds and compare against the column.
```

Comment:
```suggestion
                    // so we need to project these into single-col bounds and compare against the column.
                    // This transformation is not valid in general,
                    // however in our case it is,
                    // because we only support equality conditions for multi-column indexes.
```